conda/gateways/repodata/__init__.py

```diff
@@ -0,0 +1,283 @@
+# Copyright (C) 2012 Anaconda, Inc
+# SPDX-License-Identifier: BSD-3-Clause
+from __future__ import annotations
+
+import abc
+import bz2
+import logging
+import warnings
+from contextlib import contextmanager
+from os.path import dirname
+
+from conda.auxlib.logz import stringify
+from conda.base.constants import CONDA_HOMEPAGE_URL
+from conda.base.context import context
+from conda.common.compat import ensure_text_type
+from conda.common.url import join_url, maybe_unquote
+from conda.exceptions import (
+    CondaDependencyError,
+    CondaHTTPError,
+    CondaSSLError,
+    ProxyError,
+    UnavailableInvalidChannel,
+)
+from conda.gateways.connection import (
+    ConnectionError,
+    HTTPError,
+    InsecureRequestWarning,
+    InvalidSchema,
+    RequestsProxyError,
+    SSLError,
+)
+from conda.gateways.connection.session import CondaSession
+from conda.models.channel import Channel
+
+log = logging.getLogger(__name__)
+stderrlog = logging.getLogger("conda.stderrlog")
+
+
+class RepodataIsEmpty(UnavailableInvalidChannel):
+    """
+    Subclass used to determine when empty repodata should be cached, e.g. for a
+    channel that doesn't provide current_repodata.json
+    """
+
+    pass
+
+
+class RepoInterface(abc.ABC):
+    # TODO: Support async operations
+    # TODO: Support progress bars
+    def repodata(self, state: dict) -> str:
+        ...
+
+
+class Response304ContentUnchanged(Exception):
+    pass
+
+
+class CondaRepoInterface(RepoInterface):
+    def __init__(self, url: str, repodata_fn: str | None, **kwargs) -> None:
+        self._url = url
+        self._repodata_fn = repodata_fn
+
+        self._log = logging.getLogger(__name__)
+        self._stderrlog = logging.getLogger("conda.stderrlog")
+
+    def repodata(self, state: dict) -> str | None:
+        if not context.ssl_verify:
+            warnings.simplefilter("ignore", InsecureRequestWarning)
+
+        session = CondaSession()
+
+        headers = {}
+        etag = state.get("_etag")
+        last_modified = state.get("_mod")
+        if etag:
+            headers["If-None-Match"] = etag
+        if last_modified:
+            headers["If-Modified-Since"] = last_modified
+        filename = self._repodata_fn
+
+        url = join_url(self._url, filename)
+
+        with conda_http_errors(self._url, filename):
```

Comment:
this with statement factors out the non-reusable baroque error handling from the old function